lib/translator/write.cc

```diff
@@ -0,0 +1,217 @@
+#include "write.h"
+
+#ifdef __KERNEL__
+#include <linux/byteorder/generic.h>
+#else
+#include <netinet/in.h>
+#endif
+#include <byteswap.h>
+
+namespace translator {
+
+// anonymous namespace for helper functions and variables
+namespace {
+
+StatusCode BuildPRInfo(uint8_t wrprotect, uint8_t& pr_info) {
+  uint8_t pract, prchk;
+
+  switch (wrprotect) {
+    case 0b000:
+      pract = 1;
+      prchk = 0b000;
+      break;
+    case 0b001:
+      pract = 0;
+      prchk = 0b111;
+      break;
+    case 0b101:
+      pract = 0;
+      prchk = 0b111;
+      break;
+    case 0b010:
+      pract = 0;
+      prchk = 0b011;
+      break;
+    case 0b011:
+      pract = 0;
+      prchk = 0b00;
+      break;
+    case 0b100:
+      pract = 0;
+      prchk = 0b100;
+      break;
+    default:
+      // All other codes shall result in command termination with CHECK
+      // CONDITION status, ILLEGAL REQUEST sense key, and ILLEGAL FIELD IN CDB
+      // additional sense code.
+      DebugLog("Invalid WriteProtect Code for PRInfo.");
+      return StatusCode::kFailure;
+  }
+
+  pr_info = prchk | (pract << 3);
+  return StatusCode::kSuccess;
+}
+
+// Converts transfer_length in units of logical blocks to units of pages
+uint32_t GetTransferLengthPages(uint16_t transfer_length, uint32_t page_size,
+                                uint32_t lba_size) {
+  uint64_t transfer_length_bytes = transfer_length * lba_size;
+  return transfer_length_bytes / page_size +
```

Comment:
will page size ever be 0?

Reply:
right now the page_size is a constant that we're setting in translation.cc. Once/if we start getting page size elsewhere, then maybe we should check for non-zero values to prevent errors. But i think it's fine to leave this as is

Reply:
added a check